ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp

```diff
@@ -435,188 +638,368 @@ string get_login(string username, string password, string cap)
+string get_open_filename(string filter, string fname, string title)
 {
-  filter.append("||");
-  const unsigned int l = filter.length();
-  for (unsigned int i = 0; i < l; i++)
-    if (filter[i] == '|') filter[i] = 0;
-
-  char fn[MAX_PATH];
-  strcpy(fn, filename.c_str());
-
-  OPENFILENAME ofn;
-  ofn.lStructSize = sizeof(ofn); ofn.hwndOwner = enigma::hWnd; ofn.hInstance = NULL;
-  ofn.lpstrFilter = filter.c_str(); ofn.lpstrCustomFilter = NULL;
-  ofn.nMaxCustFilter = 0; ofn.nFilterIndex = 0;
-  ofn.lpstrFile = fn; ofn.nMaxFile = MAX_PATH;
-  ofn.lpstrFileTitle = NULL; ofn.nMaxFileTitle = 0;
-  ofn.lpstrInitialDir = NULL; ofn.lpstrTitle = caption.length() ? caption.c_str() : NULL;
-  ofn.Flags=OFN_FILEMUSTEXIST | OFN_HIDEREADONLY | OFN_NOCHANGEDIR;
-  ofn.nFileOffset = 0; ofn.nFileExtension = 0;
-  ofn.lpstrDefExt = NULL; ofn.lCustData = 0;
-  ofn.lpfnHook = NULL; ofn.lpTemplateName = 0;
-
-  bool ret = GetOpenFileName(&ofn);
-  return ret == 0 ? "-1" : fn;
-}
-
-string get_save_filename(string filter, string filename, string caption)
+  OPENFILENAMEW ofn;
+
+  string str_filter = filter.append("||");
+
+  tstring tstr_filter = widen(str_filter);
+  replace(tstr_filter.begin(), tstr_filter.end(), '|', '\0');
+  tstring tstr_fname = widen(fname);
+  tstring tstr_title = widen(title);
+  if (title == "") title = "Open";
+  gs_cap = title;
+
+  WCHAR wstr_fname[MAX_PATH + 1];
+  wcsncpy(wstr_fname, tstr_fname.c_str(), MAX_PATH + 1);
+
+  ZeroMemory(&ofn, sizeof(ofn));
+  ofn.lStructSize = sizeof(ofn);
+  ofn.hwndOwner = enigma::hWnd;
+  ofn.lpstrFile = wstr_fname;
+  ofn.nMaxFile = MAX_PATH + 1;
+  ofn.lpstrFilter = tstr_filter.c_str();
+  ofn.nFilterIndex = 0;
+  ofn.lpstrTitle = tstr_title.c_str();
+  if (title == "") ofn.lpstrTitle = NULL;
+  ofn.lpstrInitialDir = NULL;
+  ofn.Flags = OFN_EXPLORER | OFN_FILEMUSTEXIST | OFN_HIDEREADONLY;
+
+  if (GetOpenFileNameW(&ofn) != 0)
+  {
+    static string result;
+    result = shorten(wstr_fname);
+    return result;
+  }
+
+  return "";
+}
+
+string get_save_filename(string filter, string fname, string title)
+{
+  OPENFILENAMEW ofn;
+
+  string str_filter = filter.append("||");
+
+  tstring tstr_filter = widen(str_filter);
+  replace(tstr_filter.begin(), tstr_filter.end(), '|', '\0');
+  tstring tstr_fname = widen(fname);
+  tstring tstr_title = widen(title);
+  if (title == "") title = "Save As";
+  gs_cap = title;
+
+  WCHAR wstr_fname[MAX_PATH + 1];
+  wcsncpy(wstr_fname, tstr_fname.c_str(), MAX_PATH + 1);
+
+  ZeroMemory(&ofn, sizeof(ofn));
+  ofn.lStructSize = sizeof(ofn);
+  ofn.hwndOwner = enigma::hWnd;
+  ofn.lpstrFile = wstr_fname;
+  ofn.nMaxFile = MAX_PATH + 1;
+  ofn.lpstrFilter = tstr_filter.c_str();
+  ofn.nFilterIndex = 0;
+  ofn.lpstrTitle = tstr_title.c_str();
+  if (title == "") ofn.lpstrTitle = NULL;
+  ofn.lpstrInitialDir = NULL;
+  ofn.Flags = OFN_EXPLORER | OFN_PATHMUSTEXIST | OFN_HIDEREADONLY | OFN_OVERWRITEPROMPT;
```

Comment:
These 10 lines (783-793) for `OPENFILENAMEW` look virtually identical for `get_save_filename` and `get_open_filename`, please eliminate this redundancy by creating a helper method `enigma::ofw_init(...)`. Use the helper method in these two functions and then any other functions which are also duplicating this logic.